### ISSUE TITLE
Noisy `ElectronSeedProducer` in 2024 and Phase-2 worfklows 

### ISSUE THREAD

**Issue by mmusich** [cms-sw/cmssw#45658]
While investigating something else I stumbled upon the fact that `ElectronSeedProducer` is noisy in several 2024 (but also Phase2) workflows, see e.g.:
   * `13034.0`:  [log](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_14_1_X_2024-08-05-2300/pyRelValMatrixLogs/run/13034.0_TTbar_14TeV+2024PU/step3_TTbar_14TeV+2024PU.log#/)
   * `29634.0`: [log](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_14_1_X_2024-08-05-2300/pyRelValMatrixLogs/run/29634.0_TTbar_14TeV+2026D110/step3_TTbar_14TeV+2026D110.log#)
   
The warnings are of the type:

```
%MSG-w ElectronSeedGenerator:  ElectronSeedProducer:ecalDrivenElectronSeeds  06-Aug-2024 09:03:39 CEST Run: 1 Event: 303
this similar old seed already has another dRz2Pos
old seed mask/dRz2/dPhi2/dRz2Pos/dPhi2Pos: 3/0.00805616/-0.00121471/inf/inf
new seed mask/dRz2/dPhi2/dRz2Pos/dPhi2Pos: 3/inf/inf/0.00778341/-0.000662431
%MSG
```

**Comment by makortel** [comment]
assign RecoEgamma/EgammaElectronProducers

**Comment by makortel** [comment]
assign upgrade

**Comment by makortel** [comment]
FYI @cms-sw/egamma-pog-l2

**Comment by jfernan2** [comment]
@sameasy @Prasant1993 as E-gamma reco contacts, could you please have a look? Thanks

**Comment by makortel** [comment]
Any news?

**Comment by RSalvatico** [comment]
Shamik and Prasant are not EGM reco conveners anymore. Hopefully @SanghyunKo can have a look?

**Comment by jfernan2** [comment]
@RSalvatico if those persons are not EGM Reco contacts any longer, I believe we need to update the names in [1], should I add @SanghyunKo  ?

[1] https://twiki.cern.ch/twiki/bin/view/CMS/RecoContacts

**Comment by RSalvatico** [comment]
@jfernan2 I have added to the twiki the names, email addresses, and GitHub usernames of the new EGM Reco conveners

**Comment by rovere** [comment]
I think this is PR caused the triggering of so many messages:
https://github.com/cms-sw/cmssw/pull/45488

I tag @VinInn as the expert on this matter (approx_log)

**Comment by mmusich** [comment]
type egamma

**Comment by beaudett** [comment]
This PR has an other consequence, which was reported [here]( https://cms-pdmv-prod.web.cern.ch/valdb/campaigns/14_1_0_pre6_2024/report/Reconstruction.FullSim.Electron)
(The initial report was issued before the link with the PR has been established) 
Namely, the number of electron seeds made with the q<0 or q>0 assumption became different. Here is [the plot](https://cms-egamma.web.cern.ch/validation/Electrons/Releases/14_1_0_pre6_2024_DQM_std/FullvsFull_CMSSW_14_1_0_pre5/PU-PU_ZEE_14/gifs/h_ele_seedDphi2Pos.gif) showing the difference for the q>0 hypothesis. pre6 is in red, pre5 in blue. The plot is normalized, but one can see from the number of entries that we lost many seeds between the two releases.
The equivalent plot with the q<0 hypothesis is [here](https://cms-egamma.web.cern.ch/validation/Electrons/Releases/14_1_0_pre6_2024_DQM_std/FullvsFull_CMSSW_14_1_0_pre5/PU-PU_ZEE_14/gifs/h_ele_seedDphi2.gif)

I assume that the error message issued in the logs corresponds to one seed which is lost. 
Fortunately enough, it does not impact the electron efficiency. It might affect the fake rate though but it hasn't been checked. 
It has been checked (thanks @archiron and @AdrianoDee ) that if the PR is discarded one gets back the seeds, see [here](https://cms-egamma.web.cern.ch/validation/Electrons/Releases/14_1_0_pre6_2024_DQM_std/FullvsFull_CMSSW_14_1_0_pre5_b/PU-PU_ZEE_14/pngs/h_ele_seedDphi2Pos.png) (red is pre6 without the PR, blue is pre5, and we are looking at the q>0 hypothesis)

How the above PR changed the seed reconstruction is so far a mystery to me.

**Comment by makortel** [comment]
In a way the behavior prior to #45488 was undefined

**Comment by beaudett** [comment]
I am not denying that, even though I am not an expert of this code. We will try to have a deeper look at the behavior of this piece of code before and after and change.

**Comment by makortel** [comment]
So the printout
> ```
> %MSG-w ElectronSeedGenerator:  ElectronSeedProducer:ecalDrivenElectronSeeds  06-Aug-2024 09:03:39 CEST Run: 1 Event: 303
> this similar old seed already has another dRz2Pos
> old seed mask/dRz2/dPhi2/dRz2Pos/dPhi2Pos: 3/0.00805616/-0.00121471/inf/inf
> new seed mask/dRz2/dPhi2/dRz2Pos/dPhi2Pos: 3/inf/inf/0.00778341/-0.000662431
> %MSG
> ```

comes from
https://github.com/cms-sw/cmssw/blob/8340b52ad5b3272e98a300259ff0c67959520800/RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc#L73-L82

According to the printout `res.dRZPos(1)` and `res.dPhiPos(1)` are `inf`. Earlier the comparison on line 73 was against `std::numeric_limits<float>::infinity()`, and a different branch was taken.

I'd **guess** some computations result in infinities. Maybe these comparisons that were earlier against `infinity()` and are now against `max()` would need to be accompanied with `edm::isFinite()` check.

(makes me also wonder how safe use of the `fast-math` is here...)

**Comment by beaudett** [comment]
Hi 
together with @archiron we are trying to reproduce the problem. Is `fast-math` enabled as of 14_1_0_pre6 ? Is it possible to enable/disable it at the scram level ? Thanks

**Comment by makortel** [comment]
> together with @archiron we are trying to reproduce the problem. Is `fast-math` enabled as of 14_1_0_pre6 ?

Yes (I see it was enabled in https://github.com/cms-sw/cmssw/pull/13819)

> Is it possible to enable/disable it at the scram level ? Thanks

Yes, just remove or comment out the `ofast-flag` line in `BuildFile.xml`.
https://github.com/cms-sw/cmssw/blob/52219c0c2d0e662e26e724435dd8698a7b8e3b5e/RecoEgamma/EgammaElectronAlgos/BuildFile.xml#L1

**Comment by beaudett** [comment]
See this [PR](https://github.com/cms-sw/cmssw/pull/46710) from @archiron . A few `infinity()` had been forgotten in the transition. 
Running by hand on many previously verbose events, we (@archiron and I) didn't see any computation resulting in infinite (wrong calculations would probably result in `nan` anyway ?)

**Comment by hjbossi** [comment]
In order to fix this issue for 2024 HI release (14_1_X) I have backported the above PR (see here: https://github.com/cms-sw/cmssw/pull/48930). Please let me know if anything else is needed to fix this issue. Upon local tests, implemented the fix from the PR indeed suppresses the warnings.
